src/main/java/com/hubspot/jinjava/lib/filter/IntersectFilter.java

```diff
@@ -35,9 +38,41 @@ public Object filter(
     Object[] args,
     Map<String, Object> kwargs
   ) {
-    return new ArrayList<>(
-      Sets.intersection(objectToSet(var), objectToSet(parseArgs(interpreter, args)))
-    );
+    Object argObj = parseArgs(interpreter, args);
+
+    Set<Object> varSet = objectToSet(var);
+    Set<Object> argSet = objectToSet(argObj);
+
+    boolean areMismatchedElementTypes = !getTypeOfSetElements(varSet)
+      .equals(getTypeOfSetElements(argSet));
+
+    if (areMismatchedElementTypes) {
+      interpreter.addError(
+        new TemplateError(
+          TemplateError.ErrorType.WARNING,
+          TemplateError.ErrorReason.OTHER,
+          TemplateError.ErrorItem.FILTER,
+          String.format(
+            "Mismatched types. `value` elements are of type `%s` and `list` elements are of type `%s`. This may lead to unexpected behavior.",
```

Comment:
I am slightly confused by this wording. Maybe something like 
`Mismatched Types: input set has elements of type '%s' but arg set has elements of type '%s'. Use `|map` filter to convert sets to the same type for filter to work correctly.`